packages/transaction-controller/src/utils.ts

```diff
@@ -226,8 +241,122 @@ export function transactionMatchesNetwork(
   if (transaction.chainId) {
     return transaction.chainId === chainId;
   }
+
   if (transaction.networkID) {
     return transaction.networkID === networkId;
   }
   return false;
 }
+
+/**
+ * Determines the type of the transaction by analyzing the txParams.
+ * It will never return TRANSACTION_TYPE_CANCEL or TRANSACTION_TYPE_RETRY as these
+ * represent specific events that we control from the extension and are added manually
+ * at transaction creation.
+ *
+ * @param txParams - Parameters for the transaction.
+ * @param ethQuery - EthQuery instance.
+ * @returns A object with the transaction type and the contract code response in Hex.
+ */
+export async function determineTransactionType(
+  txParams: Transaction,
+  ethQuery: EthQuery,
+): Promise<InferTransactionTypeResult> {
+  let result;
+  let contractCode;
+  const { data, to } = txParams;
+  const name = parseStandardTokenTransactionData(data)?.name;
+
+  if (data && !to) {
+    result = TransactionType.deployContract;
+  } else {
+    const { contractCode: resultCode, isContractAddress } =
+      await readAddressAsContract(ethQuery, to);
+
+    contractCode = resultCode;
+
+    if (isContractAddress) {
+      const hasValue = txParams.value && Number(txParams.value) !== 0;
+
+      const tokenMethodName = [
+        TransactionType.tokenMethodApprove,
+        TransactionType.tokenMethodSetApprovalForAll,
+        TransactionType.tokenMethodTransfer,
+        TransactionType.tokenMethodTransferFrom,
+        TransactionType.tokenMethodSafeTransferFrom,
+      ].find((methodName) => methodName.toLowerCase() === name?.toLowerCase());
+
+      result =
+        data && tokenMethodName && !hasValue
+          ? tokenMethodName
+          : TransactionType.contractInteraction;
+    } else {
+      result = TransactionType.simpleSend;
+    }
+  }
+
+  return { type: result, getCodeResponse: contractCode };
+}
+
+/**
+ * Attempts to decode transaction data using ABIs for three different token standards: ERC20, ERC721, ERC1155.
+ * The data will decode correctly if the transaction is an interaction with a contract that matches one of these
+ * contract standards
+ *
+ * @param data - Encoded transaction data.
+ * @returns A representation of an ethereum contract call.
+ */
+export function parseStandardTokenTransactionData(
```

Comment:
At least for now while there is only internal usage, should we not export this and the below function to minimise the number of public methods?